p2p/host/autorelay/relay_finder.go

```diff
@@ -0,0 +1,545 @@
+				log.Debugw("skipping node. Already have enough candidates", "id", pi.ID, "num", numCandidates, "max", rf.conf.maxCandidates)
+				continue
+			}
+			rf.refCount.Add(1)
+			go func() {
+				defer rf.refCount.Done()
+				rf.handleNewNode(ctx, pi)
+			}()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (rf *relayFinder) notifyNewCandidate() {
+	select {
+	case rf.candidateFound <- struct{}{}:
+	default:
+	}
+}
+
+// handleNewNode tests if a peer supports circuit v1 or v2.
+// This method is only run on private nodes.
+// If a peer does, it is added to the candidates map.
+// Note that just supporting the protocol doesn't guarantee that we can also obtain a reservation.
+func (rf *relayFinder) handleNewNode(ctx context.Context, pi peer.AddrInfo) {
+	rf.relayMx.Lock()
+	relayInUse := rf.usingRelay(pi.ID)
+	rf.relayMx.Unlock()
+	if relayInUse {
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	defer cancel()
+	supportsV1, err := rf.tryNode(ctx, pi)
+	if err != nil {
+		log.Debugf("node %s not accepted as a candidate: %s", pi.ID, err)
+		return
+	}
+	rf.candidateMx.Lock()
+	if len(rf.candidates) > rf.conf.maxCandidates {
+		rf.candidateMx.Unlock()
+		return
+	}
+	log.Debugw("node supports relay protocol", "peer", pi.ID, "supports circuitv2", !supportsV1)
+	rf.candidates[pi.ID] = &candidate{ai: pi, isRelayV1: supportsV1}
+	rf.candidateMx.Unlock()
+
+	rf.notifyNewCandidate()
+}
+
+// tryNode checks if a peer actually supports either circuit v1 or circuit v2.
+// It does not modify any internal state.
+func (rf *relayFinder) tryNode(ctx context.Context, pi peer.AddrInfo) (supportsRelayV1 bool, err error) {
+	if err := rf.host.Connect(ctx, pi); err != nil {
+		return false, fmt.Errorf("error connecting to relay %s: %w", pi.ID, err)
+	}
+
+	conns := rf.host.Network().ConnsToPeer(pi.ID)
+	for _, conn := range conns {
+		if isRelayAddr(conn.RemoteMultiaddr()) {
+			return false, errors.New("not a public node")
+		}
+	}
+
+	// wait for identify to complete in at least one conn so that we can check the supported protocols
+	ready := make(chan struct{}, 1)
+	for _, conn := range conns {
+		go func(conn network.Conn) {
+			select {
+			case <-rf.host.IDService().IdentifyWait(conn):
+				select {
+				case ready <- struct{}{}:
+				default:
+				}
+			case <-ctx.Done():
+			}
+		}(conn)
+	}
+
+	select {
+	case <-ready:
+	case <-ctx.Done():
+		return false, ctx.Err()
+	}
+
+	protos, err := rf.host.Peerstore().SupportsProtocols(pi.ID, protoIDv1, protoIDv2)
+	if err != nil {
+		return false, fmt.Errorf("error checking relay protocol support for peer %s: %w", pi.ID, err)
+	}
+
+	// If the node speaks both, prefer circuit v2
+	var supportsV1, supportsV2 bool
+	for _, proto := range protos {
+		switch proto {
+		case protoIDv1:
+			supportsV1 = true
+		case protoIDv2:
+			supportsV2 = true
+		}
+	}
+	if !supportsV1 && !supportsV2 {
+		return false, errors.New("doesn't speak circuit v1 or v2")
+	}
+	return supportsV1, nil
```

Comment:
this will always select v1 if the node supports both, contrary to the intent.

Maybe write it as `supportsV1 && !supportsV2`